src/main/java/org/candlepin/resource/PoolResource.java

```diff
@@ -171,10 +174,18 @@ public List<PoolDTO> listPools(
         if (pageRequest != null) {
             qualifier.setOffset(pageRequest.getPage())
                 .setLimit(pageRequest.getPerPage());
+
+            boolean reverse = pageRequest.getOrder() == PageRequest.DEFAULT_ORDER;
+            qualifier.addOrder(pageRequest.getSortBy(), reverse);
         }
 
         Page<List<Pool>> pageResponse = poolManager.listAvailableEntitlementPools(qualifier);
+
         List<Pool> poolList = pageResponse.getPageData();
+        if (qualifier.getOffset() != null && qualifier.getLimit() != null) {
```

Comment:
If we have the offset and limit from the qualifier, we could be having them applied to the query rather than manually parsing out the list ourselves. Then we can just pass through the list of pools as received from the curator and throw it in the page. Unless there's something I'm not seeing that requires we pull everything and do the offset/limit ourselves.

May also be apt to not return a Page object at all from the manager/curator level. That was always a bit strange.

Reply:
As an example of this, the `AsyncJobStatusCurator` uses a `QueryArguments` object to build out queries and applies the offset+limit from it to the query. This can be seen in `.findJobs(...)` around line 307:

```
        TypedQuery<AsyncJobStatus> query = this.getEntityManager()
            .createQuery(criteriaQuery);

        if (queryArgs != null) {
            Integer offset = queryArgs.getOffset();
            if (offset != null && offset > 0) {
                query.setFirstResult(offset);
            }

            Integer limit = queryArgs.getLimit();
            if (limit != null && limit > 0) {
                query.setMaxResults(limit);
            }
        }

        return query.getResultList();
```

Reply:
I don't think we can do the paging on the query because the set of pools is being filtered again in `PoolManager.listAvailableEntitlementPools`.  If the curator returns 5 pools because the limit on the qualifier is 5 and then we filter the pools again in `PoolManager.listAvailableEntitlementPools`, then we could return less than the requested page size.

https://github.com/candlepin/candlepin/blob/jalbrech/m/CANDLEPIN-999/src/main/java/org/candlepin/controller/PoolManager.java#L1562

Reply:
Uuuuugh... alright, disregard.